crates/shared/src/gelato_api.rs

```diff
@@ -0,0 +1,210 @@
+//! Gelato HTTP API client and data types.
+//!
+//! <https://docs.gelato.network/developer-services/relay/quick-start/api>
+
+use crate::http_client::HttpClientFactory;
+use anyhow::{ensure, Result};
+use chrono::{DateTime, Utc};
+use ethcontract::{H160, H256, U256};
+use model::u256_decimal::DecimalU256;
+use reqwest::Url;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, skip_serializing_none};
+use std::{
+    env,
+    fmt::{self, Display, Formatter},
+};
+
+pub struct GelatoClient {
+    client: reqwest::Client,
+    base: Url,
+    api_key: String,
+}
+
+impl GelatoClient {
+    pub fn new(factory: &HttpClientFactory, api_key: String) -> Self {
+        Self::with_url(
+            factory,
+            Url::parse("https://relay.gelato.digital/").unwrap(),
+            api_key,
+        )
+    }
+
+    pub fn with_url(factory: &HttpClientFactory, base: Url, api_key: String) -> Self {
+        Self {
+            client: factory.create(),
+            base,
+            api_key,
+        }
+    }
+
+    pub fn from_env() -> Result<Self> {
+        Ok(Self::new(
+            &HttpClientFactory::default(),
+            env::var("GELATO_API_KEY")?,
+        ))
+    }
+
+    pub async fn sponsored_call(&self, call: &Call) -> Result<TaskId> {
+        let response = self
+            .client
+            .post(self.base.join("relays/v2/sponsored-call")?)
+            .json(&CallWithKey {
+                call,
+                sponsor_api_key: &self.api_key,
+            })
+            .send()
+            .await?;
+
+        let status = response.status();
+        let text = response.text().await?;
+        ensure!(status.is_success(), "HTTP {status} error: {text}");
+
+        let receipt = serde_json::from_str::<CallReceipt>(&text)?;
+        Ok(receipt.task_id)
+    }
+
+    pub async fn task_status(&self, id: &TaskId) -> Result<Task> {
+        let response = self
+            .client
+            .get(self.base.join("tasks/status/")?.join(&id.0)?)
+            .send()
+            .await?;
+
+        let status = response.status();
+        let text = response.text().await?;
+        ensure!(status.is_success(), "HTTP {status} error: {text}");
+
+        let data = serde_json::from_str::<TaskResponse>(&text)?;
+        Ok(data.task)
+    }
+}
+
+/// A call to send to the Gelato relay network and to be executed onchain.
+#[serde_as]
+#[skip_serializing_none]
+#[derive(Clone, Debug, Default, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Call {
+    pub chain_id: u64,
+    pub target: H160,
+    #[serde(with = "model::bytes_hex")]
+    pub data: Vec<u8>,
```

Comment:
`Debug` should probably rather print the hex string presentation here. See PR [here](https://github.com/cowprotocol/services/pull/836/files) for the required changes.